### PR TITLE
Infinitely synthetic IAM Paragraphs

### PR DESCRIPTION
**Pull request by srbhchandra**
Truly infinite IAM synthetic paragraphs dataset since every datum in every batch will be different.

Among the best runs so far: https://wandb.ai/fullstackdeeplearning/fsdl-text-recognizer-2022-training/runs/2xybxxqe
More importantly, it is 36% faster than the previous case of almost infinite IAM synthetic paragraphs where the paragraphs were randomly recreated every `n=20` epochs.

I didn't make a report specific to this run but made a report combining all the important runs at https://wandb.ai/fullstackdeeplearning/fsdl-text-recognizer-2022-training/reports/Final-set-of-all-worthwhile-runs--VmlldzoyNTM5MjU5. 

Signed-off-by: srbhchandra <srbhchandra@gmail.com>

### CODE REVIEW

File: text_recognizer/data/iam_original_and_synthetic_paragraphs.py [hunk 1]
```diff
@@ -43,18 +43,6 @@ def setup(self, stage: str = None) -> None:
...
-        self.iam_syn_paragraphs._synthesize_dataset()
-        self.data_train = ConcatDataset([self.iam_paragraphs.data_train, self.iam_syn_paragraphs.data_train])
-        return DataLoader(
-            self.data_train,
-            shuffle=True,
-            batch_size=self.batch_size,
-            num_workers=self.num_workers,
-            pin_memory=self.on_gpu,
-        )
-
```
Comment (srbhchandra) [inline text_recognizer/data/iam_original_and_synthetic_paragraphs.py:57]: removing what was added in previous PRs

File: text_recognizer/data/iam_synthetic_paragraphs.py [hunk 2]
```diff
@@ -76,13 +94,97 @@ def __repr__(self) -> str:
         )
         return basic + data
 
-    @cachedproperty
```
Comment (srbhchandra) [inline text_recognizer/data/iam_synthetic_paragraphs.py:79]: I'm not sure if caching guarantees the images will be loaded only once. Need to read more!

File: text_recognizer/data/iam_synthetic_paragraphs.py [hunk 3]
```diff
@@ -76,13 +93,97 @@ def __repr__(self) -> str:
...
-    def line_crops(self):
-        return load_processed_line_crops("train", PROCESSED_DATA_DIRNAME)
 
-    @cachedproperty
-    def line_labels(self):
-        return load_processed_line_labels("train", PROCESSED_DATA_DIRNAME)
+class IAMSyntheticParagraphsDataset(torch.utils.data.Dataset):
```
Comment (srbhchandra) [inline text_recognizer/data/iam_synthetic_paragraphs.py:97]: this probably should be in a different file?

File: training/run_experiment.py [hunk 4]
```diff
@@ -194,7 +194,7 @@ def main():
     # Hide lines below until Lab 05
     trainer.profiler = pl.profiler.PassThroughProfiler()  # turn profiling off during testing
     # Hide lines above until Lab 05
-    trainer.test(lit_model, datamodule=data)
+    trainer.test(lit_model, datamodule=data, ckpt_path="best")
```
Comment (srbhchandra) [inline training/run_experiment.py:197]: All the test metrics reported by me are with the best checkpoint.

**Comment by charlesfrye** [comment]
superseded by #50

**Pull request closed without merging**